server/src/main/java/org/elasticsearch/cluster/project/ProjectStateRegistry.java

```diff
@@ -20,28 +20,49 @@
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.set.Sets;
 import org.elasticsearch.xcontent.ToXContent;
 
 import java.io.IOException;
 import java.util.Collections;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
 
 /**
  * Represents a registry for managing and retrieving project-specific state in the cluster state.
  */
 public class ProjectStateRegistry extends AbstractNamedDiffable<ClusterState.Custom> implements ClusterState.Custom {
     public static final String TYPE = "projects_registry";
-    public static final ProjectStateRegistry EMPTY = new ProjectStateRegistry(Collections.emptyMap());
+    public static final ProjectStateRegistry EMPTY = new ProjectStateRegistry(Collections.emptyMap(), Collections.emptySet(), 0);
 
     private final Map<ProjectId, Settings> projectsSettings;
+    // Projects that have been marked for deletion based on their file-based setting
+    private final Set<ProjectId> projectsMarkedForDeletion;
+    // A counter that is incremented each time one or more projects are marked for deletion.
+    private final long projectsMarkedForDeletionGeneration;
 
     public ProjectStateRegistry(StreamInput in) throws IOException {
         projectsSettings = in.readMap(ProjectId::readFrom, Settings::readSettingsFromStream);
+        if (in.getTransportVersion().onOrAfter(TransportVersions.PROJECT_STATE_REGISTRY_RECORDS_DELETIONS)) {
+            projectsMarkedForDeletion = in.readCollectionAsSet(ProjectId::readFrom);
+            projectsMarkedForDeletionGeneration = in.readZLong();
```

Comment:
Since `projectsMarkedForDeletionGeneration` is non-negative
```suggestion
            projectsMarkedForDeletionGeneration = in.readVLong();
```